### PR TITLE
Support neovim-prompt partially to fix #38

### LINKED ISSUE #38: Can't input '9' key.
# Problems summary

I at the time of fuzzy search, can't input the '9' key.  
When I press the '9' key, the cursor will move down(it looks like 'move_to_next_line' mapping has been executed)
## Expected

That I can input '9' key.
## Environment Information
- OS:
  
  Windows10 64bit
- nvim-python-doctor result or `:CheckHealth` result(neovim ver.0.1.5-452+):
  https://github.com/tweekmonster/nvim-python-doctor
  
  Because of my vim is vim8, not neovim. So :CheckHealth command is not.
## Neovim Python Diagnostic

nothing. :version output instead.

```
VIM - Vi IMproved 8.0 (2016 Sep 12, compiled Sep 13 2016 09:59:10)
MS-Windows 64-bit GUI version
Included patches: 1-3
```
## Python versions visible to Neovim

`:py3 print('sys.version')` output is below.

```
3.5.2 (v3.5.2:4def2a2901a5, Jun 25 2016, 22:18:55) [MSC v.1900 64 bit (AMD64)]
```
### 'python' info from /usr/local/bin/nvim

nothing
### 'python3' info from /usr/local/bin/nvim

nothing.
## Python versions visible in the current shell

same is `Python versions visible to Neovim`.
## Provide a minimal init.vim with less than 50 lines (Required!)

_vimrc is below.

``` vim
set shellslash
filetype plugin indent on
let $MY_VIMRUNTIME = $VIM . "/vimfiles"

if &compatible
  set nocompatible
endif
let s:dein_dir = expand('$MY_VIMRUNTIME/dein')
let g:dein#install_log_filename = s:dein_dir . '/dein.log'
let s:dein_repo_dir = s:dein_dir . '/repos/github.com/Shougo/dein.vim'
if &runtimepath !~# '/dein.vim'
  if !isdirectory(s:dein_repo_dir)
    execute '!git clone https://github.com/Shougo/dein.vim' s:dein_repo_dir
  endif
  execute 'set runtimepath^=' . fnamemodify(s:dein_repo_dir, ':p')
endif
if dein#load_state(s:dein_dir)
  call dein#begin(s:dein_dir)
call dein#add('Shougo/dein.vim')
call dein#add('Shougo/denite.nvim')
call dein#end()
call dein#save_state()
endif
if dein#check_install()
  call dein#install()
endif

filetype plugin indent on
```
## The reproduce ways from neovim starting (Required!)
1. pre-installed python3.5, and set the PATH environment.
2. Download kaoriya-vim.  
   [Vim — KaoriYa](https://www.kaoriya.net/software/vim/)
3. put the _vimrc same location as kaoriya-vim/gvim.exe.
4. run 'gvim.exe'.
5. run ':Denite file_rec' or command, buffer, line, etc.
## Generate a logfile if appropriate

nothing.
## Screen shot (if possible)

nothing.
## Upload the log file

nothing.

### PR DESCRIPTION
**Pull request by lambdalisue**
Related #40 

While [neovim-prompt](https://github.com/lambdalisue/neovim-prompt) would become an official prompt UI of denite.nvim, I partially use the features provided by neovim-prompt to fix #38.

This PR will changes
- Custom mapping syntax. Use `'<C-a>'` instead of `"\<C-a>"`
- <del>`<S-Tab>` won't work anymore</del> -> Supported

It changed a core part of denite.nvim so **please help us with testing this branch to figure out if the current features are fully implemented or not.** Give me 👍 if it works on your environment 😄

### CODE REVIEW

**Comment by HerringtonDarkholme** [comment]
I tested it in neovim 0.1.6, python3.5, ubuntu 16.04. 
Number keys are available as expected. But shit, ctrl keys are not working properly.
`isprint` raises an error that `string of length 0 found`

**Comment by lambdalisue** [comment]
@HerringtonDarkholme I could not reproduce your issue. Could you give me a step by step procedure?

**Comment by HerringtonDarkholme** [comment]
```vimrc
set rtp+=path/todein.vim/
call dein#begin(expand('path/dein'))
call dein#add('Shougo/dein.vim')
call dein#add('Shougo/denite.nvim')
call dein#end()
filetype plugin indent on
syntax enable
```

Open nvim, Denite file_rec, press `ctrl` key.

**Comment by lambdalisue** [comment]
@HerringtonDarkholme 

I tested with the following min vimrc but could not reproduce. Did you reconstruct runtimepath? If not, try `call dein#recache_runtimepath()`

```vim
" vim -u ~/.vim/vimrc.min
if has('vim_starting')
  set nocompatible
  set runtimepath+=~/Code/github.com/lambdalisue/denite.nvim
endif

nnoremap ; :
nnoremap : ;

filetype plugin indent on
syntax on
```

**Comment by HerringtonDarkholme** [comment]
Yes, I'm sure rtp is recached. 

```viml
set rtp+=path/to/denite.nvim
filetype plugin indent on
syntax enable
```

**Comment by lambdalisue** [comment]
@HerringtonDarkholme Strange. Anyway the 86feaa0 should fix the issue.

**Comment by HerringtonDarkholme** [comment]
I print out the problematic output. It reads
```
b'<80>kb'
b'<80><fd>a'
```

**Comment by lambdalisue** [comment]
That's really strange. Do you use non 'utf-8' encoding or something like that (even so, it should be supported.)

**Comment by HerringtonDarkholme** [comment]
Yes, the new commit solve the problem. ctrl key works. but `shift-tab` is not.

I might have some idea about this strange behavior. The strange key code is provided by host terminal program, gnome-terminal gives these keys rather than \x80. When I switched to rxvt these problem went away.

**Comment by lambdalisue** [comment]
> but shift-tab is not.

As I mentioned, the feature is removed for now while my environment somehow could not produce `<S-Tab>` keycode....

> I might have some idea about this strange behavior. The strange key code is provided by host terminal program, gnome-terminal gives these keys rather than \x80. When I switched to rxvt these problem went away.

Actually, that is not strange. My iTerm does same and worked perfectly in my enviornment.
Could you try

```vim
" Check keycode {{{
function! s:getchar() abort
  redraw | echo 'Press any key: '
  let c = getchar()
  while c ==# "\<CursorHold>"
    redraw | echo 'Press any key: '
    let c = getchar()
  endwhile
  redraw | echo printf('Raw: %s | Char: %s', c, nr2char(c))
endfunction
command! GetChar call s:getchar()
" }}}
```

And `:GetChar` then press `<backspace>` on both gnome-terminal and rxvt? and the value of `&encoding`?

**Comment by HerringtonDarkholme** [comment]
Sorry I didn't notice shit's change.
On rxvt and gnome terminal, both print `<80>kb | Char:`, &encoding utf-8
But for ctrl key,  on gnome terminal prints something. rxvt does not print at all.

**Comment by lambdalisue** [comment]
> But for ctrl key, on gnome terminal prints something

Hum... It's strange while single `Ctrl` hit should not produce a meaningful keycode...

**Comment by lambdalisue** [comment]
Note: `<S-Tab>` is re-implemented.

@HerringtonDarkholme Could you tell me which key you typed to produce the followings?

```
b'<80>kb'
b'<80><fd>a'
```

**Comment by Shougo** [comment]
I will test it later.

**Comment by HerringtonDarkholme** [comment]
The new commit works like a charm. Thanks.

b'<80>kb' is backspace.
b'<80><fd>a' is ctrl.

File: doc/denite.txt [hunk 1]
```diff
@@ -96,8 +96,8 @@ EXAMPLES						*denite-examples*
 	\ ['rg', '--files'])
 
 	" Change mappings.
-	call denite#custom#map('_', "\<C-j>", 'move_to_next_line')
-	call denite#custom#map('_', "\<C-k>", 'move_to_prev_line')
+	call denite#custom#map('_', '<C-j>', 'move_to_next_line')
```
Comment (HerringtonDarkholme) [inline doc/denite.txt]: Is this doc change correct? At least I must add the backslash to make remapping work.

Comment (Shougo) [inline doc/denite.txt]: Yes, the mapping interface will be changed.

Comment (lambdalisue) [inline doc/denite.txt]: Confirmed. I'll fix

Comment (lambdalisue) [inline doc/denite.txt]: My bad. It worked on Neovim 0.1.6, Ubuntu 16.04, Gnome-terminal. Maybe the below?

1. The priority of custom '_' mapping is lower than mode mapping so `<C-j>` is override by a default mode mapping,

Comment (lambdalisue) [inline doc/denite.txt]: Probably the example on the doc is not a good example but it is not really related to this PR.

**Comment by Shougo** [comment]
I have checked the PR.  It works.
Please fix the conflict.  I will merge it.

**Comment by lambdalisue** [comment]
ok. i'll do it tonight.

**Comment by lambdalisue** [comment]
screenshot 2016-11-01 19 52 16

The above screenshot showd

1. `7` and `8` could be typed.
2. A word `denite` are highlighted by `grep` source
3. A word `FILTER` are highlighted (underlined) by prompt.

In case, could you quickly confirm if the merging was correctly performed? @Shougo, @HerringtonDarkholme

**Comment by HerringtonDarkholme** [comment]
Works like a charm! :+1:

**Comment by Shougo** [comment]
Please fix the conflict again.  I will merge it.

**Comment by lambdalisue** [comment]
Done. It should work but please confirm before merge.

**Comment by Shougo** [comment]
Merged.  Thank you!

**Comment by lambdalisue** [comment]
:tada:

**Pull request merged**